### PR TITLE
feat: make `push_str` return `self`

### LINKED ISSUE #125264: Tracking Issue for `push_str` returning `self`
Feature gate: `#![feature(push_str_returns_self)]`

This is a tracking issue for `push_str` returning `self`.

### Public API

```rust
  pub fn push_str(&mut self, string: &str) -> &mut Self {
      self.vec.extend_from_slice(string.as_bytes());
      self
  }
```

### Steps / History

- [x] Implementation: #125265
- [ ] Final comment period (FCP)[^1]
- [ ] Stabilization PR

### Unresolved Questions

- None yet.

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

### PR DESCRIPTION
**Pull request by hirehamir**
closes #125264.

### CODE REVIEW

File: library/alloc/src/rc.rs [hunk 1]
```diff
@@ -1631,7 +1631,11 @@ impl<T: ?Sized, A: Allocator> Rc<T, A> {
     #[inline]
     #[stable(feature = "rc_unique", since = "1.4.0")]
     pub fn get_mut(this: &mut Self) -> Option<&mut T> {
-        if Rc::is_unique(this) { unsafe { Some(Rc::get_mut_unchecked(this)) } } else { None }
+        if Rc::is_unique(this) {
+            unsafe { Some(Rc::get_mut_unchecked(this)) }
+        } else {
+            None
+        }
```
Comment (hamirmahal) [inline library/alloc/src/rc.rs:1638]: This change is just from formatting on save.

**Review by dtolnay — COMMENTED** [review]
I am skeptical that the benefit of changing this API is worth breaking so much code. You can see from the standard library changes in this PR, and the CI failure to build the compiler and its dependencies, that this breaks a variety of widespread patterns, including `iter.foreach(|piece| string.push_str(piece))` and `if let Some(piece) = piece { string.push_str(piece) }`.

Thanks anyway for the PR!

**Comment by hamirmahal** [comment]
You're welcome! Thanks for the rationale.

**Comment by hamirmahal** [comment]
You're welcome! Thanks for the rationale.

**Pull request closed without merging**